### PR TITLE
websockets in grpcwebproxy

### PR DESCRIPTION
**Pull request by JCGrant**
I have used these changes in a project at my job, and they seem to work well!

### CODE REVIEW

File: go/grpcwebproxy/README.md [hunk 1]
```diff
@@ -46,3 +46,14 @@ $GOPATH/bin/grpcwebproxy
     --backend_addr=localhost:9090 \
     --run_tls_server=false
 ```
+
+### Enabling Websocket Transport
+
+By default, grpcwebproxy will not use websockets as a trasport layer. To enable
```
Comment (johanbrandhorst) [inline go/grpcwebproxy/README.md]: transport

File: go/grpcwebproxy/main.go [hunk 2]
```diff
@@ -34,6 +34,8 @@ var (
 	runHttpServer = pflag.Bool("run_http_server", true, "whether to run HTTP server")
 	runTlsServer  = pflag.Bool("run_tls_server", true, "whether to run TLS server")
 
+	useWebsockets = pflag.Bool("use_websockets", false, "whether to use beta websocket tranport layer")
```
Comment (johanbrandhorst) [inline go/grpcwebproxy/main.go]: transport :thinking:

**Comment by easyCZ** [comment]
Thanks for the PR, looks good. Will test it tomorrow.

**Comment by timotheecour** [comment]
friendly ping on this

**Comment by bianbian-org** [comment]
hope merged, thanks!

**Comment by bianbian-org** [comment]
oops, websockets cannot work with  Java backend (https is ok).

**Comment by bianbian-org** [comment]
@JCGrant need help please. May I ask what is your backend server based? I use a Java-based backend, ws could return Status Code: 101 Switching Protocols, but cannot handle any other requests and disconnect when timeout. ie. proxy always got rpc error: code = Canceled:
```
grpcwsproxy_1           | time="2018-06-14T14:39:48Z" level=info msg="finished streaming call" error="rpc error: code = Canceled desc = context canceled" grpc.code=Canceled grpc.method=GetLightningDataByDateTime grpc.service=xxx.Service grpc.start_time="2018-06-14T14:38:48Z" grpc.time_ms=59971 span.kind=server system=grpc
```

**Comment by JCGrant** [comment]
@bianbian-org I use a Go backend. I'm not entirely sure what is causing your issues.

**Comment by bianbian-org** [comment]
Thanks all the same @JCGrant 
I maybe found the problem, finally!
When using grpc-web-client with [Websocket as transport](https://github.com/improbable-eng/grpc-web/blob/master/ts/src/transports/websocket.ts#L7), it depends on the finishSendFrame to tell the server that the client has finished sending:
```ts
enum WebsocketSignal {
  FINISH_SEND = 1
}

const finishSendFrame = new Uint8Array([1]);
...
    finishSend: () => {
      if (!ws || ws.readyState === ws.CONNECTING) {
        sendQueue.push(WebsocketSignal.FINISH_SEND);
      } else {
        sendToWebsocket(WebsocketSignal.FINISH_SEND);
      }
    },
```
And in the [grpc-web/websocket_wrapper](https://github.com/improbable-eng/grpc-web/blob/master/go/grpcweb/websocket_wrapper.go#L165), it waits the signal:
```go
	if len(framePayload) == 1 && framePayload[0] == 1 {
		return 0, io.EOF
	}
```

However, when call grpc methods, it do not call the finishSend(). So the grpc-web-socket-proxy is always waiting for the finishSendFrame in request till timeout. 

When add the finishSendFrame manually, it works!

**Comment by alexpdp7** [comment]
I've got this working with a Java gRPC server, sent a single request and got a response back correctly.

**Comment by jcloutz** [comment]
This works with a C# backend as well, any chance it will be merged soon?

**Pull request merged**